shipyard_blueprints/tableau/shipyard_tableau/tableau.py

```diff
@@ -1,78 +1,481 @@
-            logger.debug("No sign in method provided.")
-            try:
-                self.connect_username_password()
-                logger.authtest("Successfully connected with username_password")
-                return 0
-            except Exception as username_password_error:
-                logger.debug(
-                    f"Failed to connect with UN & PW. Message from Tableau Server {username_password_error} \n"
-                    f"Attempting with token..."
-                )
-                try:
-                    self.connect_access_token()
-                    logger.authtest("Successfully connected with Access Token")
-                    return 0
-                except Exception as token_error:
-                    logger.debug(f"Failed to connect with token: {token_error}")
-                    logger.authtest(
-                        "Authentication attempt failed.\n"
-                        "Attempted with Username & Password and Personal Access Token but could not verify "
-                        "credentials.\n"
-                        "Please ensure the account does not have MFA enabled when using Username & Password.\n"
-                        "Error messages from Tableau server:\n"
-                        f" - Username & Password error: {username_password_error}\n"
-                        f" - Personal Access Token error: {token_error}"
-                    )
-                    return 1
+    def add_user_and_pwd(self, username: str, password: str):
+        self.username = username
+        self.password = password
+        self.auth_type = "username_password"
 
-        try:
-            if sign_in_method not in ["username_password", "access_token"]:
-                logger.authtest(f"Invalid sign in method: {sign_in_method}")
-                return 1
-            elif sign_in_method == "username_password":
-                self.connect_username_password()
-
-            elif sign_in_method == "access_token":
-                self.connect_access_token()
-        except Exception as error:
-            logger.authtest(f"Could not connect to Tableau: {error}")
-            return 1
-        logger.authtest("Successfully connected to Tableau")
-        return 0
-
-    def connect_username_password(self):
+    def add_personal_access_token(
+        self, personal_access_token_name: str, personal_access_token_secret: str
+    ):
+        self.personal_access_token_name = personal_access_token_name
+        self.personal_access_token_secret = personal_access_token_secret
+        self.auth_type = "access_token"
+
+    def add_jwt(self, username: str, client_id: str, secret_id: str, secret_value: str):
+        self.username = username
+        self.client_id = client_id
+        self.secret_id = secret_id
+        self.secret_value = secret_value
+        self.auth_type = "jwt"
+
+    def connect(self, sign_in_method: str):
+        if sign_in_method not in ["username_password", "access_token", "jwt"]:
```

Comment:
Agreed that the sign_in_method will likely change to auth_type...will hold off on that for the time being. I'm going to push back against your suggestion, not because I disagree with your premise, but just the implementation. I feel that the original approach is sufficiently explicit, in fact it is catches the error up front. The readability or maintainability is not affected by this in anyway through the original implementation (IMO).